File: python_files/raylib/single_file_games/breakout.py
import pyray as p


SCREEN_WIDTH = 800
SCREEN_HEIGHT = 600
SCREEN_TITLE = "Breakout"
SCREEN_BACKGROUND = p.RAYWHITE
GAME_FPS = 60
ROWS = 5
COLS = 10
BLOCK_WIDTH = 79
BLOCK_HEIGHT = 30
BLOCK_GAP = 2


class Paddle:
    def __init__(self):
        self.width = 100
        self.height = 10
        self.x = p.get_screen_width() / 2 - self.width / 2
        self.y = p.get_screen_height() - 30
        self.color = p.BLACK
        self.speed = 5

    def draw(self):
        p.draw_rectangle_rec((self.x, self.y, self.width, self.height), self.color)

    def update(self):
        if p.is_key_down(p.KeyboardKey.KEY_LEFT) and self.x >= 0:
            self.x -= self.speed
        elif (
            p.is_key_down(p.KeyboardKey.KEY_RIGHT)
            and self.x <= p.get_screen_width() - self.width
        ):
            self.x += self.speed


class Ball:
    def __init__(self):
        self.radius = 10
        self.x = p.get_screen_width() / 2 - self.radius / 2
        self.y = p.get_screen_height() / 2 - self.radius / 2
        self.color = p.BLUE
        self.speed_x = 5
        self.speed_y = 6
        self.active = False

    def draw(self):
        p.draw_circle_v((self.x, self.y), self.radius, self.color)

    def update(self):
        if p.is_key_pressed(p.KeyboardKey.KEY_SPACE):
            self.active = True

        if self.active:
            self.x += self.speed_x
            self.y += self.speed_y

        if self.x <= 0 or self.x >= p.get_screen_width() - self.radius:
            self.speed_x *= -1
        elif self.y <= 0:
            self.speed_y *= -1


class Game:
    def __init__(self):
        self.lives = 5
        self.game_over = False
        self.game_won = False

        self.paddle = Paddle()
        self.ball = Ball()
        self.bricks = []

        for i in range(ROWS):
            for j in range(COLS):
                x = j * (BLOCK_WIDTH + BLOCK_GAP)
                y = i * (BLOCK_HEIGHT + BLOCK_GAP)

                self.bricks.append(p.Vector2(x, y))

    def draw(self):
        if not self.ball.active and not self.game_over and not self.game_won:
            p.draw_text("Press SPACE to begin!", 250, 400, 30, p.GRAY)

        # draw lives
        # p.draw_text(str(self.lives), 20, p.get_screen_height() - 30, 30, p.GRAY)
        for i in range(self.lives):
            x = 30 + i * (2 * 10 + 10)
            p.draw_circle(x, p.get_screen_height() - 30, 10, p.RED)

        self.paddle.draw()
        self.ball.draw()

        for brick in self.bricks:
            p.draw_rectangle_rec((brick.x, brick.y, BLOCK_WIDTH, BLOCK_HEIGHT), p.GRAY)

        if self.game_over:
            p.draw_text(
                "GAME OVER",
                p.get_screen_width() // 2 - 100,
                p.get_screen_height() // 2,
                40,
                p.GRAY,
            )
            p.draw_text("Press ENTER to restart!", 250, 400, 30, p.GRAY)
        if self.game_won:
            p.draw_text(
                "YOU WIN",
                p.get_screen_width() // 2 - 70,
                p.get_screen_height() // 2,
                40,
                p.GRAY,
            )
            p.draw_text("Press ENTER to restart!", 250, 400, 30, p.GRAY)

    def update(self):
        if self.lives <= 0:
            self.game_over = True

        if len(self.bricks) <= 0:
            self.game_won = True

        if not self.game_over and not self.game_won:
            self.paddle.update()
            self.ball.update()

            # ball collision paddle
            if p.check_collision_circle_rec(
                (self.ball.x, self.ball.y),
                self.ball.radius,
                (self.paddle.x, self.paddle.y, self.paddle.width, self.paddle.height),
            ):
                self.ball.speed_y *= -1

            # ball collision bricks
            for brick in self.bricks:
                if p.check_collision_circle_rec(
                    (self.ball.x, self.ball.y),
                    self.ball.radius,
                    (brick.x, brick.y, BLOCK_WIDTH, BLOCK_HEIGHT),
                ):
                    self.bricks.remove(brick)
                    self.ball.speed_y *= -1

            # update lives
            if self.ball.y >= p.get_screen_height() - self.ball.radius:
                self.lives -= 1
                self.paddle = Paddle()
                self.ball = Ball()
        else:
            if p.is_key_pressed(p.KeyboardKey.KEY_ENTER):
                self.lives = 5
                self.game_over = False
                self.game_won = False
                self.paddle = Paddle()
                self.ball = Ball()


def main():
    p.init_window(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE)
    p.set_target_fps(GAME_FPS)

    game = Game()

    while not p.window_should_close():
        p.begin_drawing()
        p.clear_background(SCREEN_BACKGROUND)

        game.draw()
        game.update()

        p.end_drawing()

    p.close_window()


if __name__ == "__main__":
    main()

File: c_files/raylib/single_file_games/breakout.c
#include "raylib.h"
#include <stdio.h>

#define LIVES_STR_LENGTH 5

#define ROWS 5
#define COLS 10
#define BLOCK_WIDTH 79
#define BLOCK_HEIGHT 30
#define BLOCK_GAP 2

// ************************************
// BRICKS
// ************************************

typedef struct Brick
{
    Vector2 pos;
    Color color;
    bool active;

} Brick;

typedef struct Bricks
{
    Brick list[ROWS * COLS];

} Bricks;

void init_bricks(Bricks *bricks)
{
    for (int i = 0; i < ROWS; i++)
    {
        for (int j = 0; j < COLS; j++)
        {
            int x = j * (BLOCK_WIDTH + BLOCK_GAP);
            int y = i * (BLOCK_HEIGHT + BLOCK_GAP);

            bricks->list[i * COLS + j].pos = (Vector2){x, y};
            bricks->list[i * COLS + j].color = GRAY;
            bricks->list[i * COLS + j].active = true;
        }
    }
}

void draw_bricks(Bricks *bricks)
{
    for (int i = 0; i < ROWS * COLS; i++)
    {
        if (bricks->list[i].active)
        {
            DrawRectangleRec((Rectangle){bricks->list[i].pos.x,
                                         bricks->list[i].pos.y,
                                         BLOCK_WIDTH,
                                         BLOCK_HEIGHT},
                             bricks->list[i].color);
        }
    }
}

// ************************************
// PADDLE
// ************************************

typedef struct Paddle
{
    float x;
    float y;
    float width;
    float height;
    Color color;
    float speed;

} Paddle;

void init_paddle(Paddle *paddle)
{
    paddle->width = 100.0f;
    paddle->height = 10.0f;
    paddle->x = (float)GetScreenWidth() / 2 - paddle->width / 2;
    paddle->y = GetScreenHeight() - 20;
    paddle->color = BLACK;
    paddle->speed = 5.0f;
}

void draw_paddle(Paddle *paddle)
{
    DrawRectangleRec((Rectangle){paddle->x,
                                 paddle->y,
                                 paddle->width,
                                 paddle->height},
                     paddle->color);
}

void update_paddle(Paddle *paddle)
{
    if (IsKeyDown(KEY_LEFT) && paddle->x >= 0)
    {
        paddle->x -= paddle->speed;
    }
    else if (IsKeyDown(KEY_RIGHT) && paddle->x <= GetScreenWidth() - paddle->width)
    {
        paddle->x += paddle->speed;
    }
}

// ************************************
// BALL
// ************************************

typedef struct Ball
{
    float x;
    float y;
    float radius;
    Color color;
    float speed_x;
    float speed_y;
    float speed_increment;
    float active;

} Ball;

void init_ball(Ball *ball)
{
    ball->x = (float)GetScreenWidth() / 2;
    ball->y = (float)GetScreenHeight() / 2;
    ball->radius = 10.0f;
    ball->color = RED;
    ball->speed_x = 4.0f;
    ball->speed_y = 4.0f;
    ball->speed_increment = 0.0005f;
    ball->active = false;
}

void draw_ball(Ball *ball)
{
    DrawCircleV((Vector2){ball->x, ball->y}, ball->radius, ball->color);
}

void update_ball(Ball *ball)
{
    if (IsKeyDown(KEY_SPACE))
    {
        ball->active = true;
    }

    if (ball->active)
    {
        ball->x += ball->speed_x;
        ball->y += ball->speed_y;

        // increment speed
        if (ball->speed_x > 0)
        {
            ball->speed_x += ball->speed_increment;
        }
        else if (ball->speed_x < 0)
        {
            ball->speed_x -= ball->speed_increment;
        }
        else if (ball->speed_y > 0)
        {
            ball->speed_y += ball->speed_increment;
        }
        else if (ball->speed_y < 0)
        {
            ball->speed_y -= ball->speed_increment;
        }
    }

    // ball collision walls
    if (ball->x <= ball->radius || ball->x >= GetScreenWidth() - ball->radius)
    {
        ball->speed_x *= -1;
    }
    if (ball->y <= ball->radius)
    {
        ball->speed_y *= -1;
    }
}

// ************************************
// MAIN LOOP
// ************************************

int main(void)
{
    const int screenWidth = 800;
    const int screenHeight = 600;
    const char screenTitle[] = "Breakout";
    const Color screenBackground = RAYWHITE;
    const int gameFps = 60;

    InitWindow(screenWidth, screenHeight, screenTitle);
    SetTargetFPS(gameFps);

    int lives = 5;
    char lives_str[LIVES_STR_LENGTH];
    bool game_over = false;
    bool game_won = false;

    Bricks bricks;
    Paddle paddle;
    Ball ball;

    init_bricks(&bricks);
    init_paddle(&paddle);
    init_ball(&ball);

    while (!WindowShouldClose())
    {
        if (!game_over && !game_won)
        {
            // game over condition
            if (lives <= 0)
            {
                game_over = true;
            }

            // game won condition
            bool all_bricks_destroyed = true;
            for (int i = 0; i < ROWS * COLS; i++)
            {
                if (bricks.list[i].active)
                {
                    all_bricks_destroyed = false;
                    break;
                }
            }
            if (all_bricks_destroyed)
            {
                game_won = true;
            }

            // convert lives to string
            snprintf(lives_str, LIVES_STR_LENGTH, "%d\n", lives);

            update_paddle(&paddle);
            update_ball(&ball);

            // update lives
            if (ball.y > GetScreenHeight())
            {
                lives--;
                init_paddle(&paddle);
                init_ball(&ball);
            }

            // ball collision paddle
            if (CheckCollisionCircleRec((Vector2){ball.x, ball.y},
                                        ball.radius,
                                        (Rectangle){paddle.x,
                                                    paddle.y,
                                                    paddle.width,
                                                    paddle.height}))
            {
                ball.speed_y *= -1;
            }

            // ball collision bricks
            for (int i = 0; i < ROWS * COLS; i++)
            {
                if (bricks.list[i].active)
                {
                    if (CheckCollisionCircleRec((Vector2){ball.x, ball.y},
                                                ball.radius,
                                                (Rectangle){bricks.list[i].pos.x,
                                                            bricks.list[i].pos.y,
                                                            BLOCK_WIDTH,
                                                            BLOCK_HEIGHT}))
                    {
                        bricks.list[i].active = false;
                        ball.speed_y *= -1;
                    }
                }
            }
        }
        else
        {
            if (IsKeyPressed(KEY_ENTER))
            {
                lives = 5;
                game_over = false;
                game_won = false;
                init_bricks(&bricks);
                init_paddle(&paddle);
                init_ball(&ball);
            }
        }

        BeginDrawing();
        ClearBackground(screenBackground);

        // game over and game won screens
        if (game_over)
        {
            DrawText("GAME OVER", GetScreenWidth() / 2 - 100, GetScreenHeight() / 2 + 100, 30, BLACK);
        }
        if (game_won)
        {
            DrawText("YOU WIN", GetScreenWidth() / 2 - 80, GetScreenHeight() / 2 + 100, 30, BLACK);
        }

        // draw lives
        DrawText(lives_str, 20, GetScreenHeight() - 40, 40, BLACK);

        draw_bricks(&bricks);
        draw_paddle(&paddle);
        draw_ball(&ball);

        EndDrawing();
    }

    CloseWindow();

    return 0;
}